org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/type/QuestionnaireValidator.java

```diff
@@ -758,6 +808,536 @@ private void validateQuestionnaireResponseItemQuantity(List<ValidationMessage> e
+          ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, precision(vdt.getValueElement().primitiveValue()) < dt, I18nConstants.QUESTIONNAIRE_QR_ITEM_DECIMAL_MAX_DECIMALS, v.primitiveValue(), dt) && ok;
+        }
+        if (qItem.hasExtension(ToolingExtensions.EXT_Q_UNIT_OPTION)) {
+          boolean matched = false;
+          List<String> allowed = new ArrayList<>();
+          for (Extension ex : qItem.getExtensionsByUrl(ToolingExtensions.EXT_Q_UNIT_OPTION)) {
+            allowed.add(gen(ex.getValueCoding()));
+            matched = matched || matchesUnit(vdt, ex.getValueCoding()); 
+          }
+          ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, matched, I18nConstants.QUESTIONNAIRE_QR_ITEM_DECIMAL_INVALID_UNITS, gen(vdt), CommaSeparatedStringBuilder.join(",", allowed)) && ok;  
+        }
+        if (qItem.hasExtension(ToolingExtensions.EXT_Q_UNIT_VALUESET)) {
+          String url = ToolingExtensions.readStringExtension(qItem, ToolingExtensions.EXT_Q_UNIT_VALUESET);
+          ValueSet vs = context.findTxResource(ValueSet.class, url);
+          if (vs == null) {
+            warning(errors, "2024-05-07", IssueType.INVARIANT, vns, false, I18nConstants.QUESTIONNAIRE_QR_ITEM_DECIMAL_INVALID_UNIT_VS, url); 
+          } else {
+            ValidationResult vr = context.validateCode(settings, vdt.getSystem(), null, vdt.getCode(), null, vs);
+            if (!vr.isOk()) {
+              ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, false, I18nConstants.QUESTIONNAIRE_QR_ITEM_DECIMAL_UNIT_FAIL_VS, gen(vdt), url) && ok;                
+            }
+          }
+        }
+      } catch (Exception e) {
+        ok = rule(errors, "2024-05-07", IssueType.STRUCTURE, vns, false, I18nConstants.QUESTIONNAIRE_QR_ITEM_DECIMAL_VALUE_EXCEPTION, e.getMessage()) && ok;
+      }
+      if (qItem.getAnswerOption().size() > 0) {
+        List<DecimalType> list = new ArrayList<DecimalType>();
+        for (QuestionnaireItemAnswerOptionComponent components : qItem.getAnswerOption()) {
+          try {
+            list.add(new DecimalType(components.getValue().primitiveValue()));
+          } catch (FHIRException e) {
+            // If it's the wrong type, just keep going
+          }
+        }
+        if (list.isEmpty()) {
+          ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, false, I18nConstants.QUESTIONNAIRE_QR_ITEM_NOOPTIONSDECIMAL) && ok;
+        } else {
+          boolean found = false;
+          for (DecimalType item : list) {
+            if (item.primitiveValue().equals(v.primitiveValue())) {
+              found = true;
+              break;
+            }
+          }
+          if (!found) {
+            ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, found, I18nConstants.QUESTIONNAIRE_QR_ITEM_NODECIMAL, v.primitiveValue()) && ok;
+          }
+        }
+      } else {
+      }
+    }
+    return ok;
+  }
+
+  private String gen(Coding c) {
```

Comment:
What does `gen` mean? Could this be a clearer name?

Reply:
I use gen() a lot to mean generate a string representation. But sure, it could be clearer